File: src/main/java/kurenai/mybot/handler/impl/AntiMiniAppHandler.java
package kurenai.mybot.handler.impl;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.xml.XmlMapper;
import kurenai.mybot.QQBotClient;
import kurenai.mybot.TelegramBotClient;
import kurenai.mybot.handler.Handler;
import lombok.extern.slf4j.Slf4j;
import net.mamoe.mirai.event.events.GroupAwareMessageEvent;
import org.apache.commons.lang3.StringUtils;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;

import java.util.Optional;

@Slf4j
@Component
@ConditionalOnProperty(prefix = "bot.handler.anti-mini-app", name = "enable", havingValue = "true", matchIfMissing = true)
@EnableConfigurationProperties(AntiMiniAppHandlerProperties.class)
public class AntiMiniAppHandler implements Handler {

    private static final String ANTI_PREFIX = "Anti mini app or share link by qq!\r\n";

    private final ObjectMapper                 jsonMapper = new ObjectMapper();
    private final XmlMapper                    xmlMapper  = new XmlMapper();
    private final AntiMiniAppHandlerProperties properties;
    private final ForwardHandlerProperties     forwardProperties;

    public AntiMiniAppHandler(AntiMiniAppHandlerProperties properties, ForwardHandlerProperties forwardProperties) {
        this.properties = properties;
        this.forwardProperties = forwardProperties;
    }

    @Override
    public boolean handle(QQBotClient client, TelegramBotClient telegramBotClient, GroupAwareMessageEvent event) {
        var id        = event.getSubject().getId();
        var chatId = forwardProperties.getGroup().getQqTelegram().getOrDefault(id, forwardProperties.getGroup().getDefaultTelegram());

        var ignoreGroup = properties.getIgnoreGroup();
        if (ignoreGroup != null && !ignoreGroup.isEmpty() && ignoreGroup.contains(id)) return true;

        var content = event.getMessage().contentToString();
        if (content.startsWith("<?xml version='1.0'")) {
            log.debug("starting parse xml: {}", content);
            try {
                var jsonNode = xmlMapper.readTree(content);
                var action   = jsonNode.get("action").asText("");
                if (action.equals("web")) {
                    var title = "title: " + Optional.ofNullable(jsonNode.get("title"))
                            .map(JsonNode::asText)
                            .or(() -> Optional.ofNullable(jsonNode.get("item"))
                                    .map(i -> i.get("summary"))
                                    .map(JsonNode::asText))
                            .orElse("");
                    var url = "url:   " + Optional.ofNullable(jsonNode.get("url")).map(JsonNode::asText).orElse("");
                    event.getSubject().sendMessage(String.join("\r\n", ANTI_PREFIX, title, url));
                    sendTg(telegramBotClient, chatId.toString(), url);
                    return false;
                }
            } catch (JsonProcessingException e) {
                log.error(e.getMessage(), e);
            }
        } else if (content.contains("\"app\":")) {
            log.debug("starting parse json: {}", content);
            try {
                var title    = "title: ";
                var url      = "url:   ";
                var jsonNode = jsonMapper.readTree(content);
                if (jsonNode.get("view").asText("").equals("news")) {
                    var news = Optional.ofNullable(jsonNode.get("meta")).map(j -> j.get("news"));
                    title += news.map(n -> n.get("title")).map(JsonNode::asText).orElse("");
                    url += news.map(n -> n.get("jumpUrl")).map(JsonNode::asText).orElse("");
                } else {
                    var item = Optional.ofNullable(jsonNode.get("meta")).map(j -> j.get("detail_1"));
                    title += item.map(m -> m.get("desc")).map(JsonNode::asText).orElse("");
                    url += item.map(m -> m.get("qqdocurl")).map(JsonNode::asText).orElse("");
                }
                event.getSubject().sendMessage(String.join("\r\n", ANTI_PREFIX, title, handleUrl(url)));
                sendTg(telegramBotClient, chatId.toString(), url);
            } catch (JsonProcessingException e) {
                log.error(e.getMessage(), e);
            }
        }
        return true;
    }

    @Override
    public int order() {
        return 50;
    }

    private void sendTg(TelegramBotClient client, String chatId, String url) {
        try {
            client.execute(SendMessage.builder()
                    .chatId(chatId)
                    .text(url)
                    .build()
            );
        } catch (TelegramApiException e) {
            log.error(e.getMessage(), e);
        }
    }

    private String handleUrl(String url) {
        if (StringUtils.isNotBlank(url) && url.contains("?") && url.contains("b23.tv")) {
            return url.substring(0, url.indexOf("?"));
        }
        return url;
    }
}

File: src/main/kotlin/kurenai/mybot/handler/impl/AntiMiniAppHandler.kt
package kurenai.mybot.handler.impl

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.xml.XmlMapper
import kurenai.mybot.handler.Handler
import kurenai.mybot.handler.config.AntiMiniAppHandlerProperties
import kurenai.mybot.handler.config.ForwardHandlerProperties
import kurenai.mybot.qq.QQBotClient
import kurenai.mybot.telegram.TelegramBotClient
import mu.KotlinLogging
import net.mamoe.mirai.event.events.GroupAwareMessageEvent
import org.apache.commons.lang3.StringUtils
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.stereotype.Component
import org.telegram.telegrambots.meta.api.methods.send.SendMessage
import org.telegram.telegrambots.meta.exceptions.TelegramApiException

@Component
@EnableConfigurationProperties(
    AntiMiniAppHandlerProperties::class, ForwardHandlerProperties::class
)
class AntiMiniAppHandler(private val properties: AntiMiniAppHandlerProperties, private val forwardProperties: ForwardHandlerProperties) : Handler {

    private val log = KotlinLogging.logger {}
    private val jsonMapper = ObjectMapper()
    private val xmlMapper = XmlMapper()

    @Throws(Exception::class)
    override suspend fun handleQQGroupMessage(client: QQBotClient, telegramBotClient: TelegramBotClient, event: GroupAwareMessageEvent): Boolean {
        val id = event.subject.id
        val chatId = forwardProperties.group.qqTelegram.getOrDefault(id, forwardProperties.group.defaultTelegram)
        val ignoreGroup = properties.ignoreGroup
        if (ignoreGroup.isNotEmpty() && ignoreGroup.contains(id)) return true
        val content = event.message.contentToString()
        val title: String
        val url: String
        if (content.startsWith("<?xml version='1.0'")) {
            log.debug("starting parse xml: $content")
            try {
                val jsonNode = xmlMapper.readTree(content)
                val action = jsonNode["action"].asText("")
                if (action == "web") {
                    title = (jsonNode["title"] ?: jsonNode["item"]?.get("summary"))?.asText() ?: ""
                    url = handleUrl(jsonNode["url"]?.asText() ?: "")
                    if (properties.enable) event.subject.sendMessage("title: $title\nurl: $url")
                    sendTg(telegramBotClient, chatId.toString(), url)
                    return false
                }
            } catch (e: JsonProcessingException) {
                log.error(e.message, e)
            }
        } else if (content.contains("\"app\":")) {
            log.debug("starting parse json: $content")
            try {
                val jsonNode = jsonMapper.readTree(content)
                if (jsonNode["view"].asText("") == "news") {
                    val news = jsonNode["meta"]?.get("news")
                    title = news?.get("title")?.asText() ?: ""
                    url = news?.get("jumpUrl")?.asText() ?: ""
                } else {
                    val item = jsonNode["meta"]?.get("detail_1")
                    title = item?.get("desc")?.asText() ?: ""
                    url = item?.get("qqdocurl")?.asText() ?: ""
                }
                handleUrl(url)
                if (properties.enable) event.subject.sendMessage("title: $title\nurl: $url")
                sendTg(telegramBotClient, chatId.toString(), url)
            } catch (e: JsonProcessingException) {
                log.error(e.message, e)
            }
        }
        return true
    }

    override fun order(): Int {
        return 50
    }

    private fun sendTg(client: TelegramBotClient, chatId: String, url: String) {
        try {
            client.execute(
                SendMessage.builder()
                    .chatId(chatId)
                    .text(url.takeIf { it.isNotEmpty() } ?: " ")
                    .build()
            )
        } catch (e: TelegramApiException) {
            log.error(e.message, e)
        }
    }

    private fun handleUrl(url: String): String {
        if (StringUtils.isNotBlank(url) && url.contains("?") && url.contains("b23.tv")) {
            // ok http get real url
            return url.substring(0, url.indexOf("?"))
        }
        return url
    }
}